src/did.ts

```diff
@@ -0,0 +1,321 @@
+    this._id = payload.did
+    return this._id
+  }
+
+  /**
+   * Create a JWS encoded signature over the given payload.
+   * Will be signed by the currently authenticated DID.
+   *
+   * @param payload             The payload to sign
+   * @param options             Optional parameters
+   */
+  async createJWS<T = any>(payload: T, options: CreateJWSOptions = {}): Promise<DagJWS> {
+    if (this._client == null) throw new Error('No provider available')
+    if (this._id == null) throw new Error('DID is not authenticated')
+    const { jws } = await this._client.request('did_createJWS', {
+      did: this._id,
+      ...options,
+      payload,
+    })
+    return jws
+  }
+
+  /**
+   * Create an IPFS compatibe DagJWS encoded signature over the given payload.
+   * Will be signed by the currently authenticated DID.
+   *
+   * @param payload             The payload to sign, may include ipld links
+   * @param options             Optional parameters
+   */
+  async createDagJWS(
+    payload: Record<string, any>,
+    options: CreateJWSOptions = {}
+  ): Promise<DagJWSResult> {
+    const { cid, linkedBlock } = await encodePayload(payload)
+    const payloadCid = encodeBase64Url(cid.bytes)
+    Object.assign(options, { linkedBlock: encodeBase64(linkedBlock) })
+    const jws = await this.createJWS(payloadCid, options)
+    jws.link = cid
+    return { jws, linkedBlock }
+  }
+
+  /**
+   * Verify a JWS. Uses the 'kid' in the header as the way to resolve
+   * the author public key.
+   *
+   * @param jws                 The JWS to verify
+   * @param atTime              JS timestamp when to verify the key
+   * @returns                   Information about the signed JWS
+   */
+  async verifyJWS(
+    jws: string | DagJWS,
+    atTime: number = new Date().valueOf()
+  ): Promise<VerifyJWSResult> {
+    if (typeof jws !== 'string') jws = fromDagJWS(jws)
+    const kid = base64urlToJSON(jws.split('.')[0]).kid as string
+    if (!kid) throw new Error('No "kid" found in jws')
+    const didResolutionResult = await this.resolve(kid)
+    const didDocumentMetadata = didResolutionResult.didDocumentMetadata
+    if (didDocumentMetadata.nextVersionId) {
+      // Maybe key is revoked
+      const nextVersion = changeVersion(kid, didDocumentMetadata.nextVersionId)
+      const nextDidResolutionResult = await this.resolve(nextVersion)
+      const verificationMethod = nextDidResolutionResult.didDocument?.verificationMethod || []
+      const keyId = parse(kid)?.fragment
+      if (keyId && !verificationMethod.find((entry) => entry.id.includes(keyId))) {
+        // Key is definitely revoked
+        const revokedAt = new Date(
+          nextDidResolutionResult.didDocumentMetadata.created || '0'
+        ).valueOf()
+        if (atTime >= revokedAt) {
+          // Do not allow using a key _after_ it is being revoked
+          throw new Error(`Key is revoked`)
+        }
+      }
+    }
```

Comment:
This seems overcomplicated and we should definitely not need to resolve the DID twice (that was why I spent a bunch of time adding this new feature to *did-core*). The suggestion should suffice!
```suggestion
    const nextUpdate = didResolutionResult.didDocumentMetadata?.nextUpdate
    if (nextUpdate) {
      // This key has been revoked. Check if the JWS 
      // was signed before it the revocation happened.
      if (atTime >= new Date(nextUpdate).valueOf()) {
       // Do not allow using a key _after_ it is being revoked
        throw new Error(`JWS was signed with a revoked key: ${kid}`)
      }
    }
```

Reply:
Is it _always_ the case, that an update to a DID document (just in case, we do not know what method is used here) results in the keys being revoked? If a DID document is updated, but it adds a key, or removes other key [and leaves the previously used signing key intact], the less-complicated design would report false failure.

Reply:
While the key *could* still be contained in the next version of the document, the `kid` that was specified is still invalid and we should consider the signature invalid imo. Otherwise for a really old version of the DID document where a key was valid until the last update we would have to resolve **all** updates inbetween the old and the last update which would be extremely inefficient.

Reply:
Okay, I consider this to be a simplification, that would work in 90% of cases.

Reply:
It should work in 100% of the cases. DID providers should make sure to always include the current version-id/time when creating JWSs

Reply:
Yes, but availability of the next version does not necessarily mean, the actual key has been revoked.

Reply:
While that's true, the assertment we are making is that the JWS was signed with the given `kid` if the DID document version is no longer valid then we should consider the JWS invalid.

Reply:
Updated the error thrown to reflect this.

Reply:
changed.